### ISSUE TITLE
Use the stdlib `os.initgroups` if it is available

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4920]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4920|
|Type|enhancement|
|Created|2011-03-01 15:20:36Z|
|Branch|https://github.com/twisted/twisted/tree/stdlib-initgroups-4920|

`os.initgroups` was added to the Python 2.7 standard library.  We should prefer it over our `twisted.python._initgroups` extension.

Attachments:

* [initgroups.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c93/c93417ca36b93214feba609e0e36d06105f4c2db/f6360978898de93c00348868bcf74473ee7e7498.patch) (579 bytes) - added by vperic on 2012-05-28 13:33:40Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4920 4920
type__enhancement enhancement
reporter__exarkun exarkun
priority__low low
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_stdlib_initgroups_4920 branches/stdlib-initgroups-4920
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1298992836000000 1298992836000000
changetime__1349271178000000 1349271178000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_1"></a>|@vperic commented|
|-|-|

As far as I can see, this is a simple matter of wrapping the current code with another "try/except" block. I've attached a patch doing just that. Not really sure how to test it, either.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve set owner to @vperic|
|-|-|

I think the tests are covering it already, and the builders on 2.7 will pick it up. But you should change twisted/topfiles/setup.py to not build _initgroups.c on 2.7 and newer. Please add a NEWS fragment as well (and merge). Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

> But you should change twisted/topfiles/setup.py to not build _initgroups.c on 2.7 and newer

With unit tests, of course.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [35884]) Branching to 'stdlib-initgroups-4920'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @itamarst<br>@exarkun set status to `new`|
|-|-|

Done and tested: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/stdlib-initgroups-4920

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst set owner to @exarkun|
|-|-|

There's a failure on one of the builds:
http://buildbot.twistedmatrix.com/builders/debian-easy-py2.6-epoll/builds/401/steps/epoll/logs/problems

Other than that, looks good; merge when fixed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [35900]) Merge stdlib-initgroups-4920 

Author: exarkun
Reviewer: itamar
Fixes: [#4920](https://github.com/twisted/twisted/issues/4920)

Use `os.initgroups` when it is available (and skip building our extension module for
the same functionality in that case).
